# Workflow 1, Module 3, Question 2

## What proteins produce agent [x]?

Let's run Q1 and use the endogenous output of that.

### Expand service

In [1]:
robokop_server = 'robokop.renci.org'

In [2]:
import requests
import pandas as pd

In [3]:
def expand(type1,identifier,type2,rebuild=None,csv=None,predicate=None):
    url=f'http://{robokop_server}:80/api/simple/expand/{type1}/{identifier}/{type2}'
    params = {'rebuild': rebuild, 
              'csv'    : csv,
              'predicate': predicate} 
    params = { k:v for k,v in params.items() if v is not None }
    response = requests.get(url,params=params)
    print( f'Return Status: {response.status_code}' )
    if response.status_code == 200:
        return response.json()
    return []

In [4]:
def parse_answer(returnanswer):
    nodes = [answer['nodes'][1] for answer in returnanswer['answers']]
    edges = [answer['edges'][0] for answer in returnanswer['answers']]
    answers = [ {"result_id": node["id"], 
                 "result_name": node["name"] if 'name' in node else node['id'], 
                 "type": edge["type"],
                 "source": edge['edge_source']}
              for node,edge in zip(nodes,edges)]
    return pd.DataFrame(answers)

### Question 2

In [8]:
FANCD2 = 'HGNC:3585'
all_results = expand('gene',FANCD2,'chemical_substance',rebuild=True)

Return Status: 200


In [30]:

CASP3='HGNC:1504'
hmm = expand('gene',CASP3,'chemical_substance',rebuild=True)

Return Status: 200


In [31]:
caspframe = parse_answer(hmm)
caspframe
#caspframe['type'].value_counts()

result_id                                        result_name  \
0              CHEBI:28834                                        CHEBI:28834   
1              CHEBI:45713                                        resveratrol   
2         PUBCHEM:58886963                                   arsenic trioxide   
3              CHEBI:29678                                    sodium arsenite   
4               CHEBI:3962                                           Curcumin   
5              CHEBI:16243                                          Quercetin   
6              CHEBI:15368                                           Acrolein   
7              CHEBI:41423                                          Celecoxib   
8              CHEBI:16469                                          Estradiol   
9              CHEBI:50924                                          sorafenib   
10               CHEBI:641                        1-Methyl-4-phenylpyridinium   
11             CHEBI:64090                            tert-Butylhydroperoxide   
12              CHEBI:4775                                       Ellagic Acid   
13                PUBCHEM:                                 aluminum maltolate   
14            MESH:D010936                                     Plant Extracts   
15            MESH:C089730                                      rosiglitazone   
16             CHEBI:90695                                    pyrazolanthrone   
17             CHEBI:35456                                   Cadmium Chloride   
18             CHEBI:37659                                     Oleanolic Acid   
19             CHEBI:17688                                           Nicotine   
20              CHEBI:9516                                       Thapsigargin   
21             CHEBI:34631                                       Chlorpyrifos   
22             CHEBI:47344                                          alvocidib   
23             CHEBI:46345                                       Fluorouracil   
24             CHEBI:34905                                           Paraquat   
25             CHEBI:88217  N,N,N',N'-tetrakis(2-pyridylmethyl)ethylenedia...   
26             CHEBI:17594                                       hydroquinone   
27              CHEBI:9150                                        Simvastatin   
28            CHEBI:136004                                        linsidomine   
29              CHEBI:4911                                          Etoposide   
...                    ...                                                ...   
1743   CHEMBL:CHEMBL101108  4-Methyl-2-(2-naphthalen-2-yl-acetylamino)-pen...   
1744  CHEMBL:CHEMBL2370213  3-Acetylamino-N-{(S)-1-[(S)-1-((S)-2-hydroxy-5...   
1745          MESH:D013859                                     Thiocarbamates   
1746   CHEMBL:CHEMBL100405  (S)-4-Methyl-2-[2-(naphthalen-2-yloxy)-acetyla...   
1747   CHEMBL:CHEMBL358725  5-{[5-(1-Carboxymethyl-2-oxo-ethylcarbamoyl)-p...   
1748   CHEMBL:CHEMBL185711  3-[(S)-2-(2,4-Difluoro-benzyloxycarbonylamino)...   
1749   CHEMBL:CHEMBL362191  (S)-5-Benzylsulfanyl-3-(2-{3-[(isoxazol-3-ylme...   
1750  CHEMBL:CHEMBL1355736                                       Theophylline   
1751       MESH:C000623387                                     HZ-6d compound   
1752            CHEBI:4821                                       ergocristine   
1753   CHEMBL:CHEMBL604054  8'-{[(2S)-2-(Methoxymethyl)pyrrolidin-1-yl]sul...   
1754   CHEMBL:CHEMBL386510  (S)-1-(4-(2-(4-methylphenylsulfonyloxy)ethoxy)...   
1755          MESH:D017967                                     Zinc Compounds   
1756   CHEMBL:CHEMBL607194  8-({(2S)-2-[(4-Chlorophenoxy)methyl]pyrrolidin...   
1757          CHEBI:133341  2,3-bis(2-hydroxyethylsulfanyl)-(1,4)naphthoqu...   
1758           CHEBI:23994                             Ethyl Methanesulfonate   
1759           CHEBI:76290    2-amino-1-methyl-6-phenylimidazo(4,5-b)pyridine   
1760           CHEBI:17847                      

In [32]:
preds=['increases_expression_of',
'increases_abundance_of',
'increases_activity_of',
'increases_stability_of',
'decreases_degradation_of',
'affects_metabolic_processing',
'affects_splicing_of',
'affects_transport_of',
'increases_response_to',
'affects_molecular_modification_of']
results = {}
for p in preds:
    print(p)
    results[p] = expand('gene',CASP3,'chemical_substance',predicate=p)

increases_expression_of
Return Status: 200
increases_abundance_of
Return Status: 200
increases_activity_of
Return Status: 200
increases_stability_of
Return Status: 200
decreases_degradation_of
Return Status: 200
affects_metabolic_processing
Return Status: 200
affects_splicing_of
Return Status: 200
affects_transport_of
Return Status: 200
increases_response_to
Return Status: 200
affects_molecular_modification_of
Return Status: 200


In [33]:
r = []
for p in preds:
    if 'answers' in results[p]:
        frame = parse_answer(results[p])
        print(p,len(frame))
        frame['predicate'] = p
        r.append(frame)
all_results = pd.concat(r)
all_results

increases_expression_of 213
increases_activity_of 684
decreases_degradation_of 7
increases_response_to 1
affects_molecular_modification_of 1


result_id                                        result_name  \
0             CHEBI:16325                                        CHEBI:16325   
1             CHEBI:32063                                        CHEBI:32063   
2             CHEBI:15738                                      Staurosporine   
3               CHEBI:641                        1-Methyl-4-phenylpyridinium   
4             CHEBI:88217  N,N,N',N'-tetrakis(2-pyridylmethyl)ethylenedia...   
5            CHEBI:136004                                        linsidomine   
6              CHEBI:8633                                           puerarin   
7             CHEBI:68508                                    diethyl maleate   
8             CHEBI:28201                                           Rotenone   
9             CHEBI:28230                                         hesperetin   
10             CHEBI:4911                                          Etoposide   
11            CHEBI:46024                                     trichostatin A   
12             CHEBI:3962                                           Curcumin   
13             CHEBI:9516                                       Thapsigargin   
14             CHEBI:4775                                       Ellagic Acid   
15            CHEBI:29678                                    sodium arsenite   
16            CHEBI:37659                                     Oleanolic Acid   
17            CHEBI:17158                                      Pyruvaldehyde   
18            CHEBI:28819                                           naringin   
19             CHEBI:5299                                         geniposide   
20            CHEBI:78492                                 diallyl trisulfide   
21            CHEBI:45713                                        resveratrol   
22            CHEBI:64090                            tert-Butylhydroperoxide   
23            CHEBI:63959                                         tripterine   
24             CHEBI:8630                                      pterostilbene   
25             CHEBI:2607                                        aloe emodin   
26            CHEBI:65408                                    andrographolide   
27            CHEBI:10136                                           gingerol   
28            CHEBI:31549                                       epothilone A   
29             CHEBI:6925                                    cyanoginosin LR   
..                    ...                                                ...   
663          MESH:C102351                                           apicidin   
664           CHEBI:30021                                   Selenomethionine   
665          MESH:D013859                                     Thiocarbamates   
666  CHEMBL:CHEMBL1355736                                       Theophylline   
667            CHEBI:4821                                       ergocristine   
668          CHEBI:133341  2,3-bis(2-hydroxyethylsulfanyl)-(1,4)naphthoqu...   
669           CHEBI:76290    2-amino-1-methyl-6-phenylimidazo(4,5-b)pyridine   
670           CHEBI:17847                                           4-cresol   
671            CHEBI:2814                                          Arecoline   
672            CHEBI:6538                                         Loratadine   
673           CHEBI:16587                                        Ecdysterone   
674           CHEBI:64310                                         crizotinib   
675           CHEBI:28217                                      Acrylonitrile   
676           CHEBI:16598                  Dichlorodiphenyl Dichloroethylene   
677           CHEBI:28747                                     picolinic acid   
678            CHEBI:5138                                        Fluvoxamine   
679           CHEBI:64318                                         Ergotamine   
680           CHEBI:15379                                             Oxygen   
681           CH

Are there any predicates indicating a production?

In [10]:
all_results['type'].value_counts()

interacts_with                            64
decreases_expression_of                   48
increases_expression_of                   27
negatively_regulates__entity_to_entity    20
positively_regulates__entity_to_entity    12
directly_interacts_with                   11
affects_expression_of                      9
increases_activity_of                      6
increases_molecular_modification_of        5
decreases_activity_of                      4
increases_metabolic_processing_of          3
increases_synthesis_of                     3
affects_response_to                        2
decreases_molecular_modification_of        1
affects_metabolic_processing_of            1
increases_degradation_of                   1
affects_activity_of                        1
Name: type, dtype: int64

In [17]:
all_results[all_results['type']=='positively_regulates__entity_to_entity']

result_id     result_name                    source  \
30   EC:1.14.13.179  EC:1.14.13.179  kegg.chemical_get_enzyme   
115    EC:2.1.1.160    EC:2.1.1.160  kegg.chemical_get_enzyme   

                                       type     Chemical  
30   positively_regulates__entity_to_entity  CHEBI:25858  
115  positively_regulates__entity_to_entity  CHEBI:27732

We're finding enzymes involved in creating the endogenous chemicals. But how come our genes are in EC format, rather than being translated to our normal HGNC?   It's because these are non-human enzymes and at the moment we're not synonymizing non-human genes. In fact, for the most part, we don't accept non-human genes, but we're not checking for that in the KEGG client.

In [11]:
all_results[all_results['type']=='increases_synthesis_of']

result_id result_name            source                    type  \
1   HGNC:2610      CYP2A6  ctd.drug_to_gene  increases_synthesis_of   
14   HGNC:435       ALOX5  ctd.drug_to_gene  increases_synthesis_of   
0     HGNC:40       ABCB1  ctd.drug_to_gene  increases_synthesis_of   

          Chemical  
1      CHEBI:68449  
14     CHEBI:28209  
0   PUBCHEM:273429

So here's an example of a metabolite (`PUBCHEM:273429`) that is increased by the action of a gene `ABC1`

### Cystic Fibrosis

In [12]:
cf = 'MONDO:0009061'
result = expand(type1,cf,type2)
result_frame = parse_answer(result)
endogenous = result_frame[ result_frame['type'] == 'related_to' ]
endogenous

Return Status: 200


result_id                   result_name  \
4            CHEBI:31835                 methylparaben   
5            CHEBI:16325              Lithocholic Acid   
6            CHEBI:30805               Dodecanoic acid   
7            CHEBI:30813                   Capric acid   
8            CHEBI:17201                Glycyl-glycine   
10           CHEBI:28621               Triethanolamine   
12           CHEBI:16755         Chenodeoxycholic Acid   
13           CHEBI:28834              Deoxycholic Acid   
14            CHEBI:6650                    Malic acid   
15           CHEBI:16359                   Cholic Acid   
27           CHEBI:16919                      Creatine   
32           CHEBI:16737                    Creatinine   
36           CHEBI:17196                  L-Asparagine   
37           CHEBI:17712                      Xanthine   
38            CHEBI:8207                   Pilocarpine   
39           CHEBI:28911                     Cobalamin   
40           CHEBI:32063                 propylparaben   
41           CHEBI:17243  Monoethylhexyl phthalic acid   
42           CHEBI:49059                 Threonic acid   
65       PUBCHEM:3952518         Prolyl-Hydroxyproline   
67        PUBCHEM:417360              Glycyl-Histidine   
69      PUBCHEM:23617285              Glycocholic acid   
72  CHEMBL:CHEMBL2004044                 Glycyl-Serine   

                        source        type  
4   hmdb.metabolite_to_disease  related_to  
5   hmdb.metabolite_to_disease  related_to  
6   hmdb.metabolite_to_disease  related_to  
7   hmdb.metabolite_to_disease  related_to  
8   hmdb.metabolite_to_disease  related_to  
10  hmdb.metabolite_to_disease  related_to  
12  hmdb.metabolite_to_disease  related_to  
13  hmdb.metabolite_to_disease  related_to  
14  hmdb.metabolite_to_disease  related_to  
15  hmdb.metabolite_to_disease  related_to  
27  hmdb.metabolite_to_disease  related_to  
32  hmdb.metabolite_to_disease  related_to  
36  hmdb.metabolite_to_disease  related_to  
37  hmdb.metabolite_to_disease  related_to  
38  hmdb.metabolite_to_disease  related_to  
39  hmdb.metabolite_to_disease  related_to  
40  hmdb.metabolite_to_disease  related_to  
41  hmdb.metabolite_to_disease  related_to  
42  hmdb.metabolite_to_disease  related_to  
65  hmdb.disease_to_metabolite  related_to  
67  hmdb.disease_to_metabolite  related_to  
69  hmdb.disease_to_metabolite  related_to  
72  hmdb.disease_to_metabolite  related_to

In [13]:
identifiers = list(endogenous['result_id'])
genes = {}
for ident in identifiers:
    print(ident)
    genes[ident] = expand('chemical_substance',ident,'gene',rebuild=True)

CHEBI:31835
Return Status: 200
CHEBI:16325
Return Status: 200
CHEBI:30805
Return Status: 200
CHEBI:30813
Return Status: 200
CHEBI:17201
Return Status: 200
CHEBI:28621
Return Status: 200
CHEBI:16755
Return Status: 200
CHEBI:28834
Return Status: 200
CHEBI:6650
Return Status: 200
CHEBI:16359
Return Status: 200
CHEBI:16919
Return Status: 200
CHEBI:16737
Return Status: 200
CHEBI:17196
Return Status: 200
CHEBI:17712
Return Status: 200
CHEBI:8207
Return Status: 200
CHEBI:28911
Return Status: 200
CHEBI:32063
Return Status: 200
CHEBI:17243
Return Status: 200
CHEBI:49059
Return Status: 200
PUBCHEM:3952518
Return Status: 200
PUBCHEM:417360
Return Status: 200
PUBCHEM:23617285
Return Status: 200
CHEMBL:CHEMBL2004044
Return Status: 200


In [14]:
results = []
for chemical in genes:
    if 'answers' in genes[chemical]:
        frame = parse_answer(genes[chemical])
        frame['Chemical'] = chemical
        results.append(frame)
all_results = pd.concat(results)
all_results['type'].value_counts()

increases_expression_of                   203
interacts_with                            119
decreases_expression_of                    87
directly_interacts_with                    43
negatively_regulates__entity_to_entity     24
positively_regulates__entity_to_entity     22
affects_response_to                        20
increases_activity_of                      18
increases_molecular_modification_of        12
affects_expression_of                       9
decreases_activity_of                       8
increases_degradation_of                    5
increases_metabolic_processing_of           2
affects_molecular_modification_of           2
decreases_abundance_of                      2
increases_secretion_of                      2
affects_localization_of                     2
increases_localization_of                   1
decreases_molecular_modification_of         1
increases_synthesis_of                      1
increases_uptake_of                         1
decreases_metabolic_processing_of 

In [15]:
all_results[all_results['type']=='positively_regulates__entity_to_entity']

result_id     result_name                    source  \
78      EC:2.8.3.25     EC:2.8.3.25  kegg.chemical_get_enzyme   
9       EC:3.1.2.21     EC:3.1.2.21  kegg.chemical_get_enzyme   
0       EC:3.5.2.13     EC:3.5.2.13  kegg.chemical_get_enzyme   
118     EC:3.5.1.24     EC:3.5.1.24  kegg.chemical_get_enzyme   
119      EC:6.2.1.7      EC:6.2.1.7  kegg.chemical_get_enzyme   
61      EC:3.5.1.24     EC:3.5.1.24  kegg.chemical_get_enzyme   
62      EC:2.8.3.25     EC:2.8.3.25  kegg.chemical_get_enzyme   
16       HGNC:15919           ACOT8  kegg.chemical_get_enzyme   
48      EC:3.5.1.24     EC:3.5.1.24  kegg.chemical_get_enzyme   
14       EC:3.9.1.1      EC:3.9.1.1  kegg.chemical_get_enzyme   
15      EC:3.5.2.10     EC:3.5.2.10  kegg.chemical_get_enzyme   
16       EC:2.1.1.2      EC:2.1.1.2  kegg.chemical_get_enzyme   
1          HGNC:753            ASNS  kegg.chemical_get_enzyme   
14       EC:6.3.1.4      EC:6.3.1.4  kegg.chemical_get_enzyme   
15       EC:6.3.1.1      EC:6.3.1.1  kegg.chemical_get_enzyme   
18      EC:3.2.2.11     EC:3.2.2.11  kegg.chemical_get_enzyme   
7        EC:3.2.2.1      EC:3.2.2.1  kegg.chemical_get_enzyme   
8    EC:1.14.13.178  EC:1.14.13.178  kegg.chemical_get_enzyme   
9    EC:1.14.13.128  EC:1.14.13.128  kegg.chemical_get_enzyme   
11   EC:1.14.13.179  EC:1.14.13.179  kegg.chemical_get_enzyme   
12       EC:3.5.4.3      EC:3.5.4.3  kegg.chemical_get_enzyme   
0       EC:3.1.1.60     EC:3.1.1.60  kegg.chemical_get_enzyme   

                                       type     Chemical  
78   positively_regulates__entity_to_entity  CHEBI:16325  
9    positively_regulates__entity_to_entity  CHEBI:30813  
0    positively_regulates__entity_to_entity  CHEBI:17201  
118  positively_regulates__entity_to_entity  CHEBI:16755  
119  positively_regulates__entity_to_entity  CHEBI:16755  
61   positively_regulates__entity_to_entity  CHEBI:28834  
62   positively_regulates__entity_to_entity  CHEBI:28834  
16   positively_regulates__entity_to_entity  CHEBI:16359  
48   positively_regulates__entity_to_entity  CHEBI:16359  
14   positively_regulates__entity_to_entity  CHEBI:16919  
15   positively_regulates__entity_to_entity  CHEBI:16919  
16   positively_regulates__entity_to_entity  CHEBI:16919  
1    positively_regulates__entity_to_entity  CHEBI:17196  
14   positively_regulates__entity_to_entity  CHEBI:17196  
15   positively_regulates__entity_to_entity  CHEBI:17196  
18   positively_regulates__entity_to_entity  CHEBI:17196  
7    positively_regulates__entity_to_entity  CHEBI:17712  
8    positively_regulates__entity_to_entity  CHEBI:17712  
9    positively_regulates__entity_to_entity  CHEBI:17712  
11   positively_regulates__entity_to_entity  CHEBI:17712  
12   positively_regulates__entity_to_entity  CHEBI:17712  
0    positively_regulates__entity_to_entity  CHEBI:17243

In [16]:
all_results[all_results['type']=='increases_synthesis_of']

result_id result_name            source                    type  \
13  HGNC:2768         DEK  ctd.drug_to_gene  increases_synthesis_of   

       Chemical  
13  CHEBI:16919

In [17]:
expand('chemical_substance',ident,'gene',predicate='increases_degradation_of')

Return Status: 200


{'message': 'Internal Server Error'}